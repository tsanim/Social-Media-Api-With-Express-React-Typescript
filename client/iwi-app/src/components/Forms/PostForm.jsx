import React from 'react';
import { wrapComponent } from 'react-snackbar-alert';
import PropTypes from 'prop-types';
import isImage from '../../utils/isImage';
import UploadedImagePreview from '../PostComponents/UploadedImagePreview';

let fd = new FormData();

export class PostForm extends React.Component {
    state = {
        text: '',
        imagePreviewUrl: '',
    }

    handleChangeInput = (e) => {
        e.persist();

        const files = e.target.files;

        if (files && files.length > 0) {
            let reader = new FileReader();
            let file = e.target.files[0];

            if (!isImage(file.type)) {
                this.props.createSnackbar({
                    message: 'You should upload only images!',
                    timeout: 3000,
                });
                return;
            }

            fd.append(e.target.name, file, e.target.name);

            reader.onload = () => {
                this.setState((oldState) => ({
                    imagePreviewUrl: reader.result,
                }));
            }

            reader.readAsDataURL(file);
        }

        this.setState({
            [e.target.name]: e.target.value,
        });
    }

    handleSubmit = (e) => {
        e.preventDefault();

        //check for empty text and send message to user or emtpy file 
        if (this.state.text === '' && !fd.get('postImg')) {
            //using createSnackbar for displaying warning message to user 
            this.props.createSnackbar({
                message: 'You can not upload post without image or text!',
                timeout: 3000,
            });
        } else {
            fd.append('text', this.state.text);
            this.props.uploadHandler(fd)
        }

        //reset state
        fd = new FormData();
        this.setState({ text: '', imagePreviewUrl: '' });
    }

    handlePreviewImgClose = (e) => {
        fd.delete('postImg');
        this.setState({ imagePreviewUrl: '' })
    }

    render() {
        const { text, imagePreviewUrl } = this.state;

        return (
            <div className="postForm">
                <form id="postForm" encType="multipart/form-data" onSubmit={this.handleSubmit}>
                    <textarea
                        name="text"
                        id="postText"
                        placeholder="Write something to your friends..."
                        value={text}
                        onChange={this.handleChangeInput}
                    />
                    <div className="inline-form-items">
                        <div className="uploadImage">
                            <label htmlFor="file"><img className="icon" src="https://cdn.iconscout.com/icon/free/png-256/gallery-44-267592.png" alt="uploadImg" />Image</label>
                            <input
                                type="file"
                                id="file"
                                name="postImg"
                                onChange={this.handleChangeInput}
                            />
                        </div>
                        <input disabled={!(fd.keys() > 0 || (text !== '' || imagePreviewUrl !== ''))} type="submit" value="Post" />
                    </div>
                </form>

                <UploadedImagePreview
                    imagePreviewUrl={imagePreviewUrl}
                    handlePreviewImgClose={this.handlePreviewImgClose}
                />
            </div>
        )
    }
}



PostForm.propTypes = {
    uploadHandler: PropTypes.func,
    createSnackbar: PropTypes.func,
}

export default wrapComponent(PostForm);