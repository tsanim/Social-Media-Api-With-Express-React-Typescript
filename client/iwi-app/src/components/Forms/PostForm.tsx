import React from 'react';
import { wrapComponent } from 'react-snackbar-alert';
import PropTypes from 'prop-types';
import isImage from '../../utils/isImage';
import UploadedImagePreview from '../PostComponents/UploadedImagePreview';
import PostFormState from '../../interfaces/Components/Forms/PostFormState.interface';
import PostFormProps from '../../interfaces/Components/Forms/PostFomProps.interface';

let fd = new FormData();

export class PostForm extends React.Component<PostFormProps, PostFormState> {
    state = {
        text: '',
        imagePreviewUrl: '',
    }

    handleChangeInput = (e: React.ChangeEvent<HTMLTextAreaElement> | React.ChangeEvent<HTMLInputElement>) => {
        e.persist();

        const files = (e.target as HTMLInputElement).files;

        if (files && files.length > 0) {
            let reader = new FileReader();
            let file = (e.target as HTMLInputElement).files[0];

            if (!isImage(file.type)) {
                this.props.createSnackbar({
                    message: 'You should upload only images!',
                    timeout: 3000,
                });
                return;
            }

            fd.append(e.target.name, file, e.target.name);

            reader.onload = () => {
                this.setState((oldState) => ({
                    imagePreviewUrl: reader.result as string,
                }));
            }

            reader.readAsDataURL(file);
        }

        this.setState({
            text: e.target.value,
        });
    }

    handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        //check for empty text and send message to user or emtpy file 
        if (this.state.text === '' && !fd.get('postImg')) {
            //using createSnackbar for displaying warning message to user 
            this.props.createSnackbar({
                message: 'You can not upload post without image or text!',
                timeout: 3000,
            });
        } else {
            fd.append('text', this.state.text);
            this.props.uploadHandler(fd);
        }

        //reset state
        fd = new FormData();
        this.setState({ text: '', imagePreviewUrl: '' });
    }

    handlePreviewImgClose = () => {
        fd.delete('postImg');
        this.setState({ imagePreviewUrl: '' })
    }

    render() {
        const { text, imagePreviewUrl } = this.state;

        return (
            <div className="postForm">
                <form id="postForm" encType="multipart/form-data" onSubmit={this.handleSubmit}>
                    <textarea
                        name="text"
                        id="postText"
                        placeholder="Write something to your friends..."
                        value={text}
                        onChange={this.handleChangeInput}
                    />
                    <div className="inline-form-items">
                        <div className="uploadImage">
                            <label htmlFor="file"><img className="icon" src="https://cdn.iconscout.com/icon/free/png-256/gallery-44-267592.png" alt="uploadImg" />Image</label>
                            <input
                                type="file"
                                id="file"
                                name="postImg"
                                onChange={this.handleChangeInput}
                            />
                        </div>
                        <input disabled={!(fd.keys() || (text !== '' || imagePreviewUrl !== ''))} type="submit" value="Post" />
                    </div>
                </form>

                <UploadedImagePreview
                    imagePreviewUrl={imagePreviewUrl}
                    handlePreviewImgClose={this.handlePreviewImgClose}
                />
            </div>
        )
    }
}

// PostForm.propTypes = {
//     uploadHandler: PropTypes.func,
//     createSnackbar: PropTypes.func,
// }

export default wrapComponent(PostForm);