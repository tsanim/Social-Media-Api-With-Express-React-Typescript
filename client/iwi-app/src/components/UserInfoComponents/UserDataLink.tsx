import React from 'react';
import { Link } from 'react-router-dom';
import URI from '../../config/config';
import PropTypes from 'prop-types';
import UserDataLinkProps from '../../interfaces/Components/UserDataComponents/UserDataLinkProps.interface';

function UserDataLink(props: UserDataLinkProps) {
    const { _id, id, imageId, username } = props.user;

    return (
        <Link to={"/profile/" + (_id ? _id : id)}>
            <img src={`${URI}/feed/image/${imageId}`} alt="userPic" />
            <figcaption>{username}</figcaption>
        </Link>
    )
}

UserDataLink.propTypes = {
    user: PropTypes.object,
}

export default UserDataLink;