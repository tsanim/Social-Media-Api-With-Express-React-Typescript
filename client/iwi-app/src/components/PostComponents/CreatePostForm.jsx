import React from 'react';
import PostForm from '../Forms/PostForm';
import UserDataLink from '../UserInfoComponents/UserDataLink';
import PropTypes from 'prop-types';

function CreatePostForm({ user, uploadHandler }) {
    return (
        <div id="makeAPost">
            <figure className="userInfo">
                <UserDataLink user={user} />
            </figure>
            <PostForm uploadHandler={uploadHandler} />
        </div>
    )
}

CreatePostForm.propTypes = {
    user: PropTypes.object,
    uploadHandler: PropTypes.func,
}

export default CreatePostForm;