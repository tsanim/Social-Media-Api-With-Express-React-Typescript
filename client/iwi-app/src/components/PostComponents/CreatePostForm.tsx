import React from 'react';
import PostForm from '../Forms/PostForm';
import UserDataLink from '../UserInfoComponents/UserDataLink';
import PropTypes from 'prop-types';
import CreatePostFormProps from '../../interfaces/Components/PostComponents/CreatePostFormProps.interface';

function CreatePostForm({ user, uploadHandler }: CreatePostFormProps) {
    return (
        <div id="makeAPost">
            <figure className="userInfo">
                <UserDataLink user={user} />
            </figure>
            <PostForm uploadHandler={uploadHandler} />
        </div>
    )
}

CreatePostForm.propTypes = {
    user: PropTypes.object,
    uploadHandler: PropTypes.func,
}

export default CreatePostForm;