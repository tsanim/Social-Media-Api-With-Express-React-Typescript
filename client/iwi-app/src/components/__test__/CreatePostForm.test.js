import React from 'react';
import CreatePostForm from '../PostComponents/CreatePostForm';

import { shallow } from 'enzyme';
import UserDataLink from '../UserInfoComponents/UserDataLink';
import PostForm from '../Forms/PostForm';

function createPostFormComponentSetup() {
    const props = {
        user: {},
        uploadHandler: jest.fn(),

    }

    const enzymeWrapper = shallow(<CreatePostForm {...props} />)

    return {
        props,
        enzymeWrapper
    }
}

describe('CreatePostForm component', () => {
    it('should render initial layout', () => {
        const { enzymeWrapper } = createPostFormComponentSetup();

        expect(enzymeWrapper.getElements()).toMatchSnapshot();
    });

    it('should render div container with id "makeAPost"', () => {
        const { enzymeWrapper } = createPostFormComponentSetup();

        expect(enzymeWrapper.find('div#makeAPost')).toHaveLength(1);
    });

    it('should render figure tag with class "userInfo"', () => {
        const { enzymeWrapper } = createPostFormComponentSetup();

        expect(enzymeWrapper.find('figure.userInfo')).toHaveLength(1);
    });

    it('should render UserDataLink Component', () => {
        const { enzymeWrapper } = createPostFormComponentSetup();

        expect(enzymeWrapper.find(UserDataLink)).toHaveLength(1);
    });

    it('should render PostForm Component', () => {
        const { enzymeWrapper } = createPostFormComponentSetup();

        expect(enzymeWrapper.find(PostForm)).toHaveLength(1);
    });
});
