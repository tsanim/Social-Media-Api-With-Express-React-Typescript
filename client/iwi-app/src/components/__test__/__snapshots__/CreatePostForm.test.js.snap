// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePostForm component should render initial layout 1`] = `
Array [
  <div
    id="makeAPost"
  >
    <figure
      className="userInfo"
    >
      <UserDataLink
        user={Object {}}
      />
    </figure>
    <Unknown
      uploadHandler={[MockFunction]}
    />
  </div>,
]
`;
